docs/python/doxygenlib/cdWriterDocstring.py

```diff
@@ -35,13 +35,82 @@
 #      This is called once the entire file has been traversed.
 #
 
+import importlib
+import os
+import re
 import sys
 import textwrap
 import types
-import re
+
+from typing import Dict
 
 from .cdUtils import *
 
+
+WORDBREAK_RE = re.compile(r"\b")
+
+MULTIPLE_WHITESPACE_RE = re.compile(r'\s+')
+MULTIPLE_WHITESPACE_REPLACE = ' '
+
+SQUARE_BRACKET_WITH_SPACE_RE = re.compile(r" *(\[|\]) *")
+SQUARE_BRACKET_WITH_SPACE_REPLACE = r"\1"
+
+COMMA_WITH_SPACE_RE = re.compile(r" *, *")
+COMMA_WITH_SPACE_REPLACE = ", "
+
+CHAR_STAR_RE = re.compile(r"\*\s*char|char\s*\*")
+CHAR_STAR_REPLACE = "str"
+
+PTR_NAME_SUFFIXES_RE = re.compile(r"(Handle|ConstPtr|RefPtr|Ptr)$")
+PTR_NAME_SUFFIXES_REPLACE = ""
+
+SMART_PTR_TEMPLATE_RE = re.compile(r"(?:shared|unique)_ptr\s*<(.*)>")
+SMART_PTR_TEMPLATE_REPLACE = r"\1"
+
+VECTOR_SUFFIX_RE = re.compile(r"\b(\w+)Vector\b")
+VECTOR_SUFFIX_REPLACE = r"list[\1]"
+
+
+def importPxrModule(moduleName: str) -> types.ModuleType:
+    """Import a single pxr submodule
+
+    moduleName is the package that follows pxr, ie, pxr.{moduleName}
+    """
+    return importlib.import_module(f"pxr.{moduleName}")
+
+def getAllPxrModules() -> Dict[str, types.ModuleType]:
+    """Find and import all pxr.* modules
+
+    Returns a dictionary mapping from submodule name to module object
+    """
+    pxrModule = importlib.import_module("pxr")
+    modules = {}
+
+    for folder in pxrModule.__path__:
+        for entry in os.scandir(folder):
+            pxrModuleName = ""
+            if entry.is_dir():
+                init_file = os.path.join(entry.path, "__init__.py")
+                if os.path.isfile(init_file):
+                    pxrModuleName = entry.name
+            elif entry.is_file():
+                if entry.name.endswith(".py"):
+                    basename = entry.name[:-len(".py")]
+                    if basename != "__init__":
+                        pxrModuleName = basename
+            if pxrModuleName:
+                modules[pxrModuleName] = importPxrModule(pxrModuleName)
+    # return a sorted dict
+    return {k: modules[k] for k in sorted(modules)}
+
+pxrModules = getAllPxrModules()
+
+# we use the reversed sorted order of the module names, because this ensures that longer names (ie, UsdGeom) come
+# before shorter ones (Usd), so the more exact matches are preferred
+PXR_MODULES_OR_JOINED = "|".join(re.escape(m) for m in reversed(pxrModules))
```

Comment:
@pmolodo in verifying this change, I'm running into an issue where reversed() isn't supported for dictionaries in Python 3.6 & 3.7 (but is supported in 3.8+).  Additionally, it sounds like dictionaries didn't really support strict ordering at all until Python 3.6 (see https://groups.google.com/g/dev-python/c/76kIoAEKlew). 

Would there be a way to make this work for earlier versions of Python 3?

Reply:
Alternatively, it'd be nice to have some sort of "requires Python 3.8 or better" check or error message. The current error you get on 3.6 is something along the lines of "Error: 'dict' object is not reversible"